### DISCUSSION TITLE
Compound variable types

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#4020] [General] [answered]
I think VB has something like this:

```
Public Sub DoStuff(ByVal obj as {SpecialObject, ISomeThing, IOtherThing})
```

This way you can constrain a variable to multiple types at once without having to actually create a class implementing all those types. This is really nice when dealing with third party libraries to which you don't have the source code!

Would be nice to have the same feature in C#:

```
public void DoStuff((SpecialObject, ISomeThing, IOtherThing) obj)
```

**Comment by CyrusNajmabadi** [comment]
Can you give an example of what you mean in VB?   Your code doesn't seem to parse.

**Comment by HaloFour** [comment]
VB.NET doesn't have union types or intersection types.  There is an existing proposal to explore them in C#, but without runtime support they would likely not work very well: https://github.com/dotnet/csharplang/discussions/399

**Comment by HaloFour** [comment, marked as answer]
You must be thinking of generic constraints, which do have a syntax like that:

```vb
Public Class C
    Public Sub M(Of T As { IDisposable, IComparable })(ByVal o As T)
        
    End Sub
End Class
```

That's a very different thing, and C# already supports that with different syntax:

```cs
public class C {
    public void M<T>(T o) where T : IDisposable, IComparable { }
}
```

**Reply by ekolis** [reply]
Oh, yes, I was looking up how to do generic constraints in VB earlier today and saw that and thought for some reason you could actually combine types like that elsewhere. Thanks!
